File: app/src/main/java/pe/andy/bookholic/searcher/Yes24LibrarySearchTask.java
package pe.andy.bookholic.searcher;

import java.io.IOException;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import pe.andy.bookholic.MainActivity;
import pe.andy.bookholic.model.Ebook;
import pe.andy.bookholic.model.SearchField;
import pe.andy.bookholic.model.SearchQuery;
import pe.andy.bookholic.model.SortBy;
import pe.andy.bookholic.util.EncodeUtil;
import pe.andy.bookholic.util.JsonParser;

public abstract class Yes24LibrarySearchTask extends LibrarySearchTask {
    String libraryName;
    String baseUrl;

    Function<Element, Ebook> ebookMapper;

    public Yes24LibrarySearchTask(MainActivity activity, String libraryName, String baseUrl) {
        super(activity, libraryName, baseUrl);

        this.libraryName = libraryName;
        this.baseUrl = baseUrl;

        this.setEncoding(Encoding_EUCKR);
    }

    @SuppressWarnings("unchecked")
    @Override
    protected String getField(SearchQuery query) {
        return SearchField.Yes24LibrarySearchField.getValue(query.getField());
    }

    String getSortBy(SearchQuery query) {
        return SortBy.Yes24LibrarySortBy.getValue(query.getSortBy());
    }

    @Override
    protected Response request(SearchQuery query) throws IOException {

        String url = this.baseUrl + "/ebook/search_list.asp";

        HttpUrl.Builder urlBuilder = HttpUrl.parse(url).newBuilder()
                .addQueryParameter("page_num", query.getPageString())
                .addQueryParameter("keyoption2", this.getField(query))
                .addQueryParameter("sort", this.getSortBy(query))
                .addEncodedQueryParameter("keyword", EncodeUtil.toEuckr(query.getKeyword()));

        OkHttpClient client = new OkHttpClient();
        Request req = new Request.Builder()
                .url(urlBuilder.build().toString())
                .addHeader("accept", "text/html; charset=euc-kr")
                .addHeader("user-agent", userAgent)
                .build();

        Response resp = client.newCall(req).execute();
        return resp;
    }

    @Override
    protected List<Ebook> parse(String html) throws IOException {
        Document doc = Jsoup.parse(html);
        Elements elems = doc.select("div.sub_main_total");

        // meta info
        Element h2 = elems.select("h2").first();
        this.resultCount = JsonParser.parseOnlyIntFrom(h2, "strong");
        this.resultPageCount = JsonParser.parseOnlyIntFrom(h2, "span.normal", "(.*-| /)");

        // no result
        if (this.resultCount == 0)
            return new ArrayList<>();

        // content
        elems = elems.select("ul");

        List<Ebook> ebooks = elems.stream()
                .map(getEbookMapper())
                .collect(Collectors.toList());

        return ebooks;
    }

    public abstract Function<Element, Ebook> getEbookMapper();
}

File: app/src/main/java/pe/andy/bookholic/searcher/Yes24LibrarySearchTask.kt
package pe.andy.bookholic.searcher

import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import pe.andy.bookholic.fragment.SearchFragment
import pe.andy.bookholic.model.Ebook
import pe.andy.bookholic.model.Library
import pe.andy.bookholic.model.LibraryType
import pe.andy.bookholic.model.SearchField.Yes24LibrarySearchField.Companion.getValue
import pe.andy.bookholic.model.SearchQuery
import pe.andy.bookholic.model.SortBy.Yes24LibrarySortBy.Companion.getValue
import pe.andy.bookholic.parser.Yes24TypeAParser
import pe.andy.bookholic.parser.Yes24TypeBParser
import pe.andy.bookholic.util.EncodingUtil.toEuckr
import pe.andy.bookholic.util.HttpExtension
import pe.andy.bookholic.util.SslTrust
import pe.andy.bookholic.util.StringExtension
import java.io.IOException
import java.lang.ref.SoftReference

class Yes24LibrarySearchTask(
        override var searchFragment: SearchFragment,
        library: Library
) : LibrarySearchTask(searchFragment, library),
        StringExtension, HttpExtension {

    init {
        this.encoding = library.encoding
    }

    override fun getLibraryCode() = library.code

    override fun create(): LibrarySearchTask {
        return SoftReference<LibrarySearchTask>(
                Yes24LibrarySearchTask(searchFragment, library)
        ).get()!!
    }

    override fun getField(query: SearchQuery): String {
        return getValue(query.field)
    }

    fun getSortBy(query: SearchQuery): String {
        return getValue(query.sortBy)
    }

    @Throws(IOException::class)
    override fun request(query: SearchQuery): Response {
        val req: Request = Request.Builder()
                .url(getUrl(query))
                .accept(library.encoding)
                .userAgent(userAgent)
                .build()

        return SslTrust.trustAllSslClient(OkHttpClient())
                .newCall(req)
                .execute()
    }

    private fun getUrl(query: SearchQuery): String {
        return when(library.type) {
            LibraryType.Yes24_A -> getUrlTypeA(query)
            LibraryType.Yes24_B -> getUrlTypeB(query)
            else -> ""
        }
    }

    private fun getUrlTypeA(query: SearchQuery): String {
        val url = "${library.url}/ebook/search_list.asp"
        return url.toHttpUrlOrNull()!!.newBuilder()
            .addQueryParameters(
                mapOf(
                    "page_num" to query.page.toString(),
                    "keyoption2" to getField(query),
                    "sort" to getSortBy(query)
                )
            )
                .addEncodedQueryParameter("keyword", toEuckr(query.keyword))
                .build()
                .toString()
    }

    private fun getUrlTypeB(query: SearchQuery): String {
        val url = "${library.url}/search/"

        val keyword = when(library.encoding) {
            Encoding_EUCKR -> query.keyword.encodeToEucKR()
            else -> query.keyword
        }

        return url.toHttpUrlOrNull()!!.newBuilder()
            .addQueryParameters(
                mapOf(
                    "page_num" to query.page.toString(),
                    "srch_order" to "total"
                )
            )
                .addEncodedQueryParameter("src_key", keyword)
                .build()
                .toString()
    }

    @Throws(IOException::class)
    override fun parse(html: String): List<Ebook> {
        val doc: Document = Jsoup.parse(html)

        parseMetaCount(doc)
        return parseBook(doc)
    }

    private fun parseMetaCount(doc: Document) {
        val pair = when(library.type) {
            LibraryType.Yes24_A -> Yes24TypeAParser.parseMetaCount(doc, library)
            LibraryType.Yes24_B -> Yes24TypeBParser.parseMetaCount(doc, library)
            else -> (0 to 0)
        }
        resultCount = pair.first
        resultPageCount = pair.second
    }

    private fun parseBook(doc: Document): List<Ebook> {

        if (resultCount == 0)
            return emptyList()

        return when(library.type) {
            LibraryType.Yes24_A -> {
                doc.select("div.sub_main_total ul")
                        .asSequence()
                        .map { Yes24TypeAParser.parse(it, library = library) }
                        .toList()
            }
            LibraryType.Yes24_B -> {
                doc.select("#booklist .line")
                        .asSequence()
                        .map { Yes24TypeBParser.parse(it, library = library) }
                        .toList()
            }
            else -> emptyList()
        }
    }
}